<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="theme-color" content="#222"><meta name="generator" content="Hexo 5.4.2">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">



<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.1.1/css/all.min.css" integrity="sha256-DfWjNxDkM94fVBWx1H5BMMp0Zq7luBlV8QRcSES7s+0=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/animate.css/3.1.1/animate.min.css" integrity="sha256-PR7ttpcvz8qrF57fur/yAx1qXMFJeJFiA6pSzWi0OIE=" crossorigin="anonymous">

<script class="next-config" data-name="main" type="application/json">{"hostname":"example.com","root":"/","images":"/images","scheme":"Muse","darkmode":false,"version":"8.11.1","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12},"copycode":{"enable":false,"style":null},"bookmark":{"enable":false,"color":"#222","save":"auto"},"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"stickytabs":false,"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"fadeInDown","post_body":"fadeInDown","coll_header":"fadeInLeft","sidebar":"fadeInUp"}},"prism":false,"i18n":{"placeholder":"Searching...","empty":"We didn't find any results for the search: ${query}","hits_time":"${hits} results found in ${time} ms","hits":"${hits} results found"}}</script><script src="/js/config.js"></script>

    <meta name="description" content="1    是什么定义：异常是指程序在执行过程中出现的非正常情况，最终会导致程序的非正常停止。 【异常并不是语法错误，语法错误的话编译时就会直接报错，不会产生字节码文件，根本不能运行。】  在 Java 等面向对象的编程语言中，异常本身就是一个类，产生异常就是创建异常对象并抛出了一个异常对象。 Java处理异常的方式是中断处理。  2    异常的体系结构    Throwable 是类，不是接口">
<meta property="og:type" content="article">
<meta property="og:title" content="异常">
<meta property="og:url" content="http://example.com/2021/07/31/%E5%BC%82%E5%B8%B8/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="1    是什么定义：异常是指程序在执行过程中出现的非正常情况，最终会导致程序的非正常停止。 【异常并不是语法错误，语法错误的话编译时就会直接报错，不会产生字节码文件，根本不能运行。】  在 Java 等面向对象的编程语言中，异常本身就是一个类，产生异常就是创建异常对象并抛出了一个异常对象。 Java处理异常的方式是中断处理。  2    异常的体系结构    Throwable 是类，不是接口">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="http://example.com/2021/07/31/%E5%BC%82%E5%B8%B8/%E5%BC%82%E5%B8%B8/Throwable.png">
<meta property="og:image" content="http://example.com/2021/07/31/%E5%BC%82%E5%B8%B8/%E5%BC%82%E5%B8%B8/image-20210730163920638.png">
<meta property="og:image" content="http://example.com/2021/07/31/%E5%BC%82%E5%B8%B8/%E5%BC%82%E5%B8%B8/image-20210730163330302.png">
<meta property="article:published_time" content="2021-07-31T02:35:42.000Z">
<meta property="article:modified_time" content="2021-07-31T04:34:29.159Z">
<meta property="article:author" content="John Doe">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/2021/07/31/%E5%BC%82%E5%B8%B8/%E5%BC%82%E5%B8%B8/Throwable.png">


<link rel="canonical" href="http://example.com/2021/07/31/%E5%BC%82%E5%B8%B8/">



<script class="next-config" data-name="page" type="application/json">{"sidebar":"","isHome":false,"isPost":true,"lang":"en","comments":true,"permalink":"http://example.com/2021/07/31/%E5%BC%82%E5%B8%B8/","path":"2021/07/31/异常/","title":"异常"}</script>

<script class="next-config" data-name="calendar" type="application/json">""</script>
<title>异常 | Hexo</title>
  





  <noscript>
    <link rel="stylesheet" href="/css/noscript.css">
  </noscript>
</head>

<body itemscope itemtype="http://schema.org/WebPage" class="use-motion">
  <div class="headband"></div>

  <main class="main">
    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar" role="button">
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <i class="logo-line"></i>
      <p class="site-title">Hexo</p>
      <i class="logo-line"></i>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>







</div>
        
  
  <div class="toggle sidebar-toggle" role="button">
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
  </div>

  <aside class="sidebar">

    <div class="sidebar-inner sidebar-nav-active sidebar-toc-active">
      <ul class="sidebar-nav">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <div class="sidebar-panel-container">
        <!--noindex-->
        <div class="post-toc-wrap sidebar-panel">
            <div class="post-toc animated"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#1-%E6%98%AF%E4%BB%80%E4%B9%88"><span class="nav-number">1.</span> <span class="nav-text">1    是什么</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-%E5%BC%82%E5%B8%B8%E7%9A%84%E4%BD%93%E7%B3%BB%E7%BB%93%E6%9E%84"><span class="nav-number">2.</span> <span class="nav-text">2    异常的体系结构</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Error"><span class="nav-number">2.1.</span> <span class="nav-text">Error</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Exception"><span class="nav-number">2.2.</span> <span class="nav-text">Exception</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#%E7%BC%96%E8%AF%91%E6%97%B6%E5%BC%82%E5%B8%B8"><span class="nav-number">2.2.1.</span> <span class="nav-text">编译时异常</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E8%BF%90%E8%A1%8C%E6%97%B6%E5%BC%82%E5%B8%B8"><span class="nav-number">2.2.2.</span> <span class="nav-text">运行时异常</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E6%80%BB%E7%BB%93"><span class="nav-number">2.2.3.</span> <span class="nav-text">总结</span></a></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#3-Throwable%E7%B1%BB%E5%B8%B8%E7%94%A8%E6%96%B9%E6%B3%95"><span class="nav-number">3.</span> <span class="nav-text">3    Throwable类常用方法</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#4-%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E6%96%B9%E5%BC%8F"><span class="nav-number">4.</span> <span class="nav-text">4    异常处理方式</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E6%96%B9%E5%BC%8F%E4%B8%80%EF%BC%9Atry-catch-finally"><span class="nav-number">4.1.</span> <span class="nav-text">异常处理方式一：try-catch-finally</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E6%96%B9%E5%BC%8F%E4%BA%8C%EF%BC%9Athrows"><span class="nav-number">4.2.</span> <span class="nav-text">异常处理方式二：throws</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%AF%B9%E6%AF%94%E4%B8%A4%E7%A7%8D%E5%A4%84%E7%90%86%E6%96%B9%E5%BC%8F"><span class="nav-number">4.3.</span> <span class="nav-text">对比两种处理方式</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E4%BD%93%E4%BC%9A%E5%BC%80%E5%8F%91%E4%B8%AD%E5%BA%94%E8%AF%A5%E5%A6%82%E4%BD%95%E9%80%89%E6%8B%A9%E4%B8%A4%E7%A7%8D%E5%A4%84%E7%90%86%E6%96%B9%E5%BC%8F%EF%BC%9F"><span class="nav-number">4.4.</span> <span class="nav-text">体会开发中应该如何选择两种处理方式？</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#5-%E6%89%8B%E5%8A%A8%E6%8A%9B%E5%87%BA%E5%BC%82%E5%B8%B8%E5%AF%B9%E8%B1%A1"><span class="nav-number">5.</span> <span class="nav-text">5    手动抛出异常对象</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#6-%E8%87%AA%E5%AE%9A%E4%B9%89%E5%BC%82%E5%B8%B8%E7%B1%BB"><span class="nav-number">6.</span> <span class="nav-text">6    自定义异常类</span></a></li></ol></div>
        </div>
        <!--/noindex-->

        <div class="site-overview-wrap sidebar-panel">
          <div class="site-author site-overview-item animated" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">John Doe</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap site-overview-item animated">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
        <a href="/archives/">
          <span class="site-state-item-count">224</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
          <a href="/categories/">
        <span class="site-state-item-count">61</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
          <a href="/tags/">
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>



        </div>
      </div>
    </div>
  </aside>
  <div class="sidebar-dimmer"></div>


    </header>

    
  <div class="back-to-top" role="button" aria-label="Back to top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>

<noscript>
  <div class="noscript-warning">Theme NexT works best with JavaScript enabled</div>
</noscript>


    <div class="main-inner post posts-expand">


  


<div class="post-block">
  
  

  <article itemscope itemtype="http://schema.org/Article" class="post-content" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/07/31/%E5%BC%82%E5%B8%B8/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="John Doe">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Hexo">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="post" itemscope itemtype="http://schema.org/CreativeWork">
      <meta itemprop="name" content="异常 | Hexo">
      <meta itemprop="description" content="">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          异常
        </h1>

        <div class="post-meta-container">
          <div class="post-meta">
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar"></i>
      </span>
      <span class="post-meta-item-text">Posted on</span>
      

      <time title="Created: 2021-07-31 10:35:42 / Modified: 12:34:29" itemprop="dateCreated datePublished" datetime="2021-07-31T10:35:42+08:00">2021-07-31</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-folder"></i>
      </span>
      <span class="post-meta-item-text">In</span>
        <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
          <a href="/categories/JavaSE/" itemprop="url" rel="index"><span itemprop="name">JavaSE</span></a>
        </span>
    </span>

  
</div>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">
        <h2 id="1-是什么"><a href="#1-是什么" class="headerlink" title="1    是什么"></a>1    是什么</h2><p><strong>定义</strong>：异常是指程序在执行过程中出现的非正常情况，最终会导致程序的非正常停止。</p>
<p>【异常并不是语法错误，语法错误的话编译时就会直接报错，不会产生字节码文件，根本不能运行。】</p>
<ul>
<li>在 Java 等面向对象的编程语言中，异常本身就是一个类，产生异常就是创建异常对象并抛出了一个异常对象。</li>
<li>Java处理异常的方式是<strong>中断处理</strong>。</li>
</ul>
<h2 id="2-异常的体系结构"><a href="#2-异常的体系结构" class="headerlink" title="2    异常的体系结构"></a>2    异常的体系结构</h2><blockquote>
<ul>
<li>  Throwable 是<strong>类</strong>，不是接口</li>
<li>  同样的，Error 和 Exception 继承了 Throwable ，所以它俩也是<strong>类</strong></li>
<li>  RuntimeException 继承了 Exception，也是一个<strong>类</strong>；</li>
<li>  但是 CheckedException 并没有真正地继承 Exception 类，并且它是一个<strong>接口</strong>。</li>
</ul>
</blockquote>
<p><img src="%E5%BC%82%E5%B8%B8/Throwable.png"></p>
<p>​        在 Java 中，所有的异常都有一个共同的祖先 <strong>java.lang.Throwable</strong> 类。Throwable 类有两个重要的子类 <strong>Exception</strong>（异常） 和 <strong>Error</strong>（错误） ，二者都是 Java 异常处理的重要子类，各自都包含大量子类。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">* java.lang.Throwable</span><br><span class="line">* 		|-----java.lang.Error:一般不编写针对性的代码进行处理。</span><br><span class="line">* 		|-----java.lang.Exception:可以进行异常的处理</span><br><span class="line">* 			|------编译时异常(checked)</span><br><span class="line">* 					|-----IOException</span><br><span class="line">* 						|-----FileNotFoundException</span><br><span class="line">* 					|-----ClassNotFoundException</span><br><span class="line">* 			|------运行时异常(unchecked,RuntimeException)</span><br><span class="line">* 					|-----NullPointerException</span><br><span class="line">* 					|-----ArrayIndexOutOfBoundsException</span><br><span class="line">* 					|-----ClassCastException</span><br><span class="line">* 					|-----NumberFormatException</span><br><span class="line">* 					|-----InputMismatchException</span><br><span class="line">* 					|-----ArithmeticException</span><br></pre></td></tr></table></figure>





<h3 id="Error"><a href="#Error" class="headerlink" title="Error"></a>Error</h3><blockquote>
<p>  <strong>java.lang.Error</strong></p>
</blockquote>
<p><strong>Error 错误是程序无法处理的错误</strong>。大多数是 JVM 层面的问题。即代码在逻辑上是正确的，但是由于 JVM 内存不足或者某些对象申请的内存空间过大导致等原因导致程序出现错误。 这些错误是在编译期是不可查的，因为它们在应用程序的控制和处理能力之外。</p>
<p>​        例如，Java虚拟机运行错误（Virtual MachineError），当 JVM 不再有继续执行操作所需的内存资源时，将出现 <code>OutOfMemoryError</code>。这些异常发生时，Java虚拟机（JVM）一般会选择线程终止。</p>
<h3 id="Exception"><a href="#Exception" class="headerlink" title="Exception"></a>Exception</h3><blockquote>
<p>  <strong>java.lang.Exception</strong></p>
</blockquote>
<p><strong>Exception 是程序本身可以处理的异常</strong>。Exception 又可分为 <strong>编译时异常</strong> 与 <strong>运行时异常</strong> 两大类。</p>
<p><img src="%E5%BC%82%E5%B8%B8/image-20210730163920638.png" alt="image-20210730163920638"></p>
<ul>
<li>  编译时异常：执行 <code>javac.exe</code> 命名时，可能出现的异常；</li>
<li>  运行时异常：执行 <code>java.exe</code> 命名时，出现的异常；</li>
</ul>
<h4 id="编译时异常"><a href="#编译时异常" class="headerlink" title="编译时异常"></a>编译时异常</h4><blockquote>
<p>  <strong>执行 <code>javac.exe</code> 命名时，可能出现的异常；</strong></p>
</blockquote>
<p>以下是一个编译时异常的简单例子。由于我们使用的是 IDEA 集成开发工具，代码是即时编译的，所以可以编译时异常不会在控制台上输出，而是非常智能的提示出来了。但是如果我们使用基本的文本工具进行开发，编写的 <code>.java</code> 代码是不会自动提示错误的，只有在经过 javac 编译后才会提示，这就是所谓的<strong>编译时异常</strong>。</p>
<img src="%E5%BC%82%E5%B8%B8/image-20210730163330302.png" alt="image-20210730163330302" style="zoom: 67%;" />



<h4 id="运行时异常"><a href="#运行时异常" class="headerlink" title="运行时异常"></a>运行时异常</h4><blockquote>
<p>  <strong>执行 <code>java.exe</code> 命名时，出现的异常；</strong></p>
</blockquote>
<p>程序在执行 <code>java.exe</code> 运行后提示的异常。</p>
<blockquote>
<p>  常见的运行时异常</p>
</blockquote>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">NullPointerException	<span class="comment">// 要访问的变量没有引用任何对象时，抛出该异常）、</span></span><br><span class="line">ArithmeticException		<span class="comment">// 算术运算异常，一个整数除以0时，抛出该异常）和 </span></span><br><span class="line">ArrayIndexOutOfBoundsException	<span class="comment">// 下标越界异常</span></span><br></pre></td></tr></table></figure>





<h4 id="总结"><a href="#总结" class="headerlink" title="总结"></a>总结</h4><p><strong>异常和错误的区别</strong>：异常能被程序本身可以处理，错误是无法处理。</p>
<p><strong>如何理解</strong></p>
<p>我们可以看到，无论是编译期异常还是运行期异常，产生异常后程序员可以在代码层面纠正，使程序能继续正常运行。如上面的编译期异常，可以通过 <code>try...catch</code> 捕获处理；对于运行时期的各种异常，如空指针异常，可以通过加判断条件来避免。这些异常都是代码层面可以处理的。而对于一些虚拟机的异常，主要是涉及到了虚拟机的配置，可能出现在某写配置较好的虚拟机上可以正常运行，而在性能较差的虚拟机上就不能运行了的情况，所以说 Error 是无法避免的。</p>
<h2 id="3-Throwable类常用方法"><a href="#3-Throwable类常用方法" class="headerlink" title="3    Throwable类常用方法"></a>3    Throwable类常用方法</h2><table>
<thead>
<tr>
<th align="center">ID</th>
<th align="center">方法</th>
<th align="center">返回值</th>
<th align="center">说明</th>
</tr>
</thead>
<tbody><tr>
<td align="center">1</td>
<td align="center"><code>printStackTrace()</code></td>
<td align="center"><code>void</code></td>
<td align="center">在控制台上打印异常的详细信息</td>
</tr>
<tr>
<td align="center">2</td>
<td align="center"><code>getMessage()</code></td>
<td align="center"><code>String</code></td>
<td align="center">获取发生异常的原因</td>
</tr>
<tr>
<td align="center">3</td>
<td align="center"><code>toString</code></td>
<td align="center"><code>String</code></td>
<td align="center">获取异常的类型和详细信息</td>
</tr>
</tbody></table>
<hr>
<h2 id="4-异常处理方式"><a href="#4-异常处理方式" class="headerlink" title="4    异常处理方式"></a>4    异常处理方式</h2><p>异常的两种处理方式：1、<strong>“抛”</strong>  2、<strong>“抓”</strong></p>
<blockquote>
<ol>
<li> <strong>“抛”</strong> </li>
</ol>
</blockquote>
<ul>
<li>  程序在正常执行的过程中，一旦出现异常，就会在异常代码处产生一个对应异常类的对象，并将此对象抛出。</li>
<li>  一旦抛出异常对象以后，其后的代码就不再执行。</li>
</ul>
<p><strong>关于异常对象的产生：</strong></p>
<ol>
<li> 系统自动生成的异常对象；</li>
<li> 手动的生成一个异常对象，并抛出（throw）。</li>
</ol>
<blockquote>
<ol start="2">
<li> <strong>“抓”</strong></li>
</ol>
</blockquote>
<p>可以理解为异常的处理方式：①<code>try-catch-finally</code>  ②<code>throws</code></p>
<h3 id="异常处理方式一：try-catch-finally"><a href="#异常处理方式一：try-catch-finally" class="headerlink" title="异常处理方式一：try-catch-finally"></a>异常处理方式一：try-catch-finally</h3><blockquote>
<p>  <strong>使用说明：</strong></p>
</blockquote>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span></span><br><span class="line">&#123;</span><br><span class="line">    <span class="comment">// 可能出现异常的代码。</span></span><br><span class="line">    <span class="comment">// 代码中出现的异常可能不止一种，针对不同类型异常可以使用多个catch语句分别做不同的处理 </span></span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">catch</span>(异常类型<span class="number">1</span> 变量名<span class="number">1</span>)</span><br><span class="line">&#123;</span><br><span class="line">    <span class="comment">// 处理异常的方式1</span></span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">catch</span>(异常类型<span class="number">2</span> 变量名<span class="number">2</span>)</span><br><span class="line">&#123;</span><br><span class="line">    <span class="comment">// 处理异常的方式2</span></span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">catch</span>(异常类型<span class="number">3</span> 变量名<span class="number">3</span>)</span><br><span class="line">&#123;</span><br><span class="line">    <span class="comment">// 处理异常的方式3</span></span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">finally</span>&#123;</span><br><span class="line">    <span class="comment">// 一定会执行的代码</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><strong>说明：</strong></p>
<ol>
<li><code>finally</code> 是可选的。</li>
<li>使用 <code>try</code> 将可能出现异常代码包起来，在执行过程中，一旦出现异常，就会生成一个对应异常类的对象，根据此异常对象的类型，去 <code>catch</code> 中进行匹配；</li>
<li>一旦 <code>try</code> 中的异常对象匹配到某一个 <code>catch</code> 时，就进入 <code>catch</code> 中进行异常的处理。一旦处理完成，就跳出当前的 <code>try-catch</code> 结构，继续执行 <code>try</code> 中后续的代码，最后再执行 <code>finally</code> 中的代码。</li>
<li><code>catch</code> 中的异常类型如果没子父类关系，则谁声明在上，谁声明在下无所谓。但是 <strong><code>catch</code> 中的异常类型如果满足子父类关系，则要求子类异常类型一定要声明在父类异常类型的上面</strong>，否则报错。【因为 <code>catch</code> 是从上向下匹配的，如果父类异常类型声明在先，则该父类异常的所有子类异常都会被父类异常拦截（多态），根本执行不到子类异常处 】</li>
<li>常用的异常对象处理的方式： <code>① String  getMessage()    ② printStackTrace()</code></li>
<li>在 <code>try</code> 结构中声明的变量，再出了 <code>try</code> 结构以后，就不能再被调用。</li>
<li><code>try-catch-finally</code> 结构可以嵌套。</li>
</ol>
<blockquote>
<p>  <strong>如何看待代码中的编译时异常和运行时异常？</strong></p>
</blockquote>
<ol>
<li> 使用 <code>try-catch-finally</code> 处理编译时异常，使得程序在编译时就不再报错，但是运行时仍可能报错。相当于我们使用 <code>try-catch-finally</code> 将一个编译时可能出现的异常，延迟到运行时出现。</li>
<li> 开发中，由于运行时异常比较常见，所以我们通常就不针对运行时异常编写 <code>try-catch-finally</code> 了。针对于编译时异常，我们则一定要考虑异常的处理。</li>
</ol>
<h3 id="异常处理方式二：throws"><a href="#异常处理方式二：throws" class="headerlink" title="异常处理方式二：throws"></a>异常处理方式二：throws</h3><ul>
<li>  <strong>throws + 异常类型</strong> 写在方法的声明处。指明此方法执行时，可能会抛出的异常类型。</li>
<li>  一旦当方法体执行时出现异常，仍会在异常代码处生成一个异常类的对象，此对象满足 <code>throws</code> 抛出的异常类型时，就会被抛出。异常代码后续的代码，就不再执行。</li>
</ul>
<h3 id="对比两种处理方式"><a href="#对比两种处理方式" class="headerlink" title="对比两种处理方式"></a>对比两种处理方式</h3><ul>
<li>  <code>try-catch-finally</code>：真正的将异常给处理掉了。</li>
<li>  <code>throws</code>：只是将异常抛给了方法的调用者，并没真正将异常处理掉。</li>
</ul>
<h3 id="体会开发中应该如何选择两种处理方式？"><a href="#体会开发中应该如何选择两种处理方式？" class="headerlink" title="体会开发中应该如何选择两种处理方式？"></a>体会开发中应该如何选择两种处理方式？</h3><ul>
<li>  如果父类中被重写的方法没 <code>throws</code> 方式处理异常，则子类重写的方法也不能使用 <code>throws</code>。这意味着如果子类重写的方法中异常，必须使用 <code>try-catch-finally</code> 方式处理。</li>
<li>  执行的方法 <code>methodA()</code> 中，先后又调用了另外的几个方法，这几个方法是递进关系执行的。我们建议这几个方法使用 <code>throws</code> 的方式进行处理，而执行的方法 <code>methodA()</code> 可以考虑使用 <code>try-catch-finally</code> 方式进行处理。</li>
<li>  使用 <code>throws</code> 抛出的异常最外层还是要落实到使用 <code>try-catch-finally</code> 处理。</li>
</ul>
<blockquote>
<p>  <strong>补充：</strong></p>
</blockquote>
<ul>
<li>  方法重写的规则之一：子类重写的方法抛出的异常类型不大于父类被重写的方法抛出的异常类型。【因为父类是子类的顶级抽象，如果子类的异常类型范围大于父类类型的话，则父类就会无法处理超出范围的异常类型】</li>
</ul>
<hr>
<h2 id="5-手动抛出异常对象"><a href="#5-手动抛出异常对象" class="headerlink" title="5    手动抛出异常对象"></a>5    手动抛出异常对象</h2><blockquote>
<ol>
<li> <strong>使用说明</strong></li>
</ol>
</blockquote>
<p>在程序执行中，除了自动抛出异常对象的情况之外，我们还可以通过 <code>throw</code> 手动抛出一个异常类的对象。</p>
<blockquote>
<ol start="2">
<li> <strong>throw 和  throws 的区别？</strong></li>
</ol>
</blockquote>
<p><strong>throw</strong>：</p>
<ul>
<li>  动词，表示抛出一个异常类的对象。</li>
<li>  生成异常对象的过程，声明在方法体内。</li>
<li>  <code>throw</code> 抛出的对象会由 <code>throws</code> 接收并抛给上一层，并在最外层由 <code>try-catch-finally</code> 处理，或直接中断程序。</li>
</ul>
<p><strong>throws</strong>：</p>
<ul>
<li>  名词，属于异常处理的一种方式。声明在方法的声明处。</li>
<li>  <code>throws</code> 用于接收程序运行时由系统抛出或者用户自定义抛出的异常，并交给 <code>try-catch-finally</code> 处理。  </li>
</ul>
<p><strong>例子：</strong></p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">Student</span>&#123;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">private</span> <span class="type">int</span> id;</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">void</span> <span class="title function_">regist</span><span class="params">(<span class="type">int</span> id)</span> <span class="keyword">throws</span> Exception &#123;</span><br><span class="line">        <span class="keyword">if</span>(id &gt; <span class="number">0</span>)&#123;</span><br><span class="line">            <span class="built_in">this</span>.id = id;</span><br><span class="line">        &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">            <span class="comment">// 手动抛出异常对象</span></span><br><span class="line"><span class="comment">//			throw new RuntimeException(&quot;您输入的数据非法！&quot;);</span></span><br><span class="line"><span class="comment">//			throw new Exception(&quot;您输入的数据非法！&quot;);</span></span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">MyException</span>(<span class="string">&quot;不能输入负数&quot;</span>);	<span class="comment">// 自定义异常对象</span></span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>





<h2 id="6-自定义异常类"><a href="#6-自定义异常类" class="headerlink" title="6    自定义异常类"></a>6    自定义异常类</h2><blockquote>
<p>  <a target="_blank" rel="noopener" href="https://lvnengdong.github.io/2020/11/26/%E9%80%9A%E7%94%A8%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E5%99%A8/">超链接：开发中的通用异常处理</a></p>
</blockquote>
<blockquote>
<p>  <strong>如何自定义一个异常类？</strong></p>
</blockquote>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">/*</span></span><br><span class="line"><span class="comment"> * 如何自定义异常类？</span></span><br><span class="line"><span class="comment"> * 1. 继承于现的异常结构：RuntimeException 、Exception</span></span><br><span class="line"><span class="comment"> * 2. 提供全局常量：serialVersionUID</span></span><br><span class="line"><span class="comment"> * 3. 提供重载的构造器</span></span><br><span class="line"><span class="comment"> * </span></span><br><span class="line"><span class="comment"> */</span></span><br><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">MyException</span> <span class="keyword">extends</span> <span class="title class_">Exception</span>&#123;</span><br><span class="line">	</span><br><span class="line">	<span class="keyword">static</span> <span class="keyword">final</span> <span class="type">long</span> <span class="variable">serialVersionUID</span> <span class="operator">=</span> -<span class="number">7034897193246939L</span>;</span><br><span class="line">	</span><br><span class="line">	<span class="keyword">public</span> <span class="title function_">MyException</span><span class="params">()</span>&#123;</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	<span class="keyword">public</span> <span class="title function_">MyException</span><span class="params">(String msg)</span>&#123;</span><br><span class="line">		<span class="built_in">super</span>(msg);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>


    </div>

    
    
    

    <footer class="post-footer">

        

          <div class="post-nav">
            <div class="post-nav-item">
                <a href="/2021/07/30/%E7%BA%A2%E9%BB%91%E6%A0%91/" rel="prev" title="红黑树">
                  <i class="fa fa-chevron-left"></i> 红黑树
                </a>
            </div>
            <div class="post-nav-item">
                <a href="/2021/08/02/%E6%B3%9B%E5%9E%8B/" rel="next" title="泛型">
                  泛型 <i class="fa fa-chevron-right"></i>
                </a>
            </div>
          </div>
    </footer>
  </article>
</div>






</div>
  </main>

  <footer class="footer">
    <div class="footer-inner">


<div class="copyright">
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">John Doe</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.js.org/muse/" rel="noopener" target="_blank">NexT.Muse</a>
  </div>

    </div>
  </footer>

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/animejs/3.2.1/anime.min.js" integrity="sha256-XL2inqUJaslATFnHdJOi9GfQ60on8Wx1C2H8DYiN1xY=" crossorigin="anonymous"></script>
<script src="/js/comments.js"></script><script src="/js/utils.js"></script><script src="/js/motion.js"></script><script src="/js/schemes/muse.js"></script><script src="/js/next-boot.js"></script>

  





  





</body>
</html>
